core/src/main/java/org/apache/iceberg/deletes/TargetedPositionDeleteWriter.java

```diff
@@ -0,0 +1,133 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.CharSequenceSet;
+
+/**
+ * A position delete writer that produces a separate delete file for each referenced data file.
+ *
+ * <p>This writer does not keep track of seen deletes and assumes all incoming records are ordered
+ * by file and position as required by the spec. If there is no external process to order the
+ * records, consider using {@link SortingPositionOnlyDeleteWriter} instead.
+ */
+public class TargetedPositionDeleteWriter<T>
+    implements FileWriter<PositionDelete<T>, DeleteWriteResult> {
+
+  private final Supplier<FileWriter<PositionDelete<T>, DeleteWriteResult>> writers;
+  private final List<DeleteFile> deleteFiles;
+  private final CharSequenceSet referencedDataFiles;
+
+  private FileWriter<PositionDelete<T>, DeleteWriteResult> currentWriter = null;
+  private CharSequence currentPath = null;
+  private boolean closed = false;
+
+  public TargetedPositionDeleteWriter(
+      Supplier<FileWriter<PositionDelete<T>, DeleteWriteResult>> writers) {
+    this.writers = writers;
+    this.deleteFiles = Lists.newArrayList();
+    this.referencedDataFiles = CharSequenceSet.empty();
+  }
+
+  @Override
+  public void write(PositionDelete<T> positionDelete) {
+    writer(positionDelete.path()).write(positionDelete);
+  }
+
+  private FileWriter<PositionDelete<T>, DeleteWriteResult> writer(CharSequence path) {
+    if (currentWriter == null) {
+      openCurrentWriter(path);
+    } else if (unequal(currentPath, path)) {
+      closeCurrentWriter();
+      openCurrentWriter(path);
+    }
+
+    return currentWriter;
+  }
+
+  @Override
+  public long length() {
+    throw new UnsupportedOperationException(getClass().getName() + " does not implement length");
+  }
+
+  @Override
+  public DeleteWriteResult result() {
+    Preconditions.checkState(closed, "Cannot get result from unclosed writer");
+    return new DeleteWriteResult(deleteFiles, referencedDataFiles);
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (!closed) {
+      closeCurrentWriter();
+      this.closed = true;
+    }
+  }
+
+  private void openCurrentWriter(CharSequence path) {
+    Preconditions.checkState(!closed, "Writer has already been closed");
+    this.currentWriter = writers.get();
+    this.currentPath = path;
+  }
+
+  private void closeCurrentWriter() {
+    if (currentWriter != null) {
+      try {
+        currentWriter.close();
+        DeleteWriteResult result = currentWriter.result();
+        deleteFiles.addAll(result.deleteFiles());
+        referencedDataFiles.addAll(result.referencedDataFiles());
+        this.currentWriter = null;
+        this.currentPath = null;
+      } catch (IOException e) {
+        throw new UncheckedIOException("Failed to close current writer", e);
+      }
+    }
+  }
+
+  private static boolean unequal(CharSequence s1, CharSequence s2) {
```

Comment:
It would be nice to put this in `CharSeqComparator` for reuse. I think it is fine that it doesn't worry about high surrogates.

Reply:
Our `CharSeqComparator` is private. I've put this into a new utility class.